src/pages/RecipeEditor.js

```diff
@@ -0,0 +1,533 @@
+                      <Form.Label>Batch Amount</Form.Label>
+                      <InputGroup>
+                        <Form.Control
+                          name="desiredVolume"
+                          type="number"
+                          value={desiredVolume}
+                          onChange={this.handleUserInput}
+                          placeholder="0"
+                          required
+                        />
+                        <InputGroup.Append>
+                          <InputGroup.Text id="amount-unit">mL</InputGroup.Text>
+                        </InputGroup.Append>
+                        <Form.Control.Feedback type="invalid">
+                          Please provide a batch amount.
+                        </Form.Control.Feedback>
+                      </InputGroup>
+                    </Form.Group>
+                  </Form.Row>
+                  <Form.Row>
+                    <Form.Group as={Col} md="4" controlId="nicotineStrength">
+                      <Form.Label>Base Strength</Form.Label>
+                      <InputGroup>
+                        <Form.Control
+                          name="nicotineStrength"
+                          type="number"
+                          value={nicotineStrength}
+                          onChange={this.handleUserInput}
+                          placeholder="0"
+                          required
+                        />
+                        <InputGroup.Append>
+                          <InputGroup.Text id="amount-unit">
+                            mg/mL
+                          </InputGroup.Text>
+                        </InputGroup.Append>
+                      </InputGroup>
+                    </Form.Group>
+                    <Form.Group as={Col} md="8">
+                      <Form.Label>Nicotine VG/PG ratio</Form.Label>
+                      <SplitSlider
+                        name="nicotineDiluentRatio"
+                        initialValue={nicotineDiluentRatio * 100}
+                        onChange={this.handleUserInput}
+                      />
+                    </Form.Group>
+                  </Form.Row>
+                  <Form.Row>
+                    <Form.Group
+                      as={Col}
+                      md="4"
+                      controlId="desiredNicotineStrength"
+                    >
+                      <Form.Label>Desired Strength</Form.Label>
+                      <InputGroup>
+                        <Form.Control
+                          name="desiredNicotineStrength"
+                          type="number"
+                          value={desiredNicotineStrength}
+                          onChange={this.handleUserInput}
+                          placeholder="0"
+                          required
+                        />
+                        <InputGroup.Append>
+                          <InputGroup.Text id="amount-unit">
+                            mg/mL
```

Comment:
I've seen `ml` other places, should keep it consistent. I tend to prefer mL, for what it's worth.

Reply:
OK, I think I got them all.